File: 백준/Platinum/3197. 백조의 호수/백조의 호수.cc
#include <bits/stdc++.h>
using namespace std;
const int dx[] = { -1, 0, 1, 0 };
const int dy[] = { 0, 1, 0, -1 };
int r, c;
char a[1504][1504];
int visitedSwan[1504][1504], visitedWater[1504][1504];
queue<pair<int, int>> waterQ, waterTempQ, swanQ, swanTempQ;
pair<int, int> swan;
int day;

void Qclear(queue<pair<int, int>>& q) {
	queue<pair<int, int>> empty;
	swap(q, empty);
}

bool moveSwan() {
	while (swanQ.size()) {
		int x = swanQ.front().first;
		int y = swanQ.front().second;
		swanQ.pop();

		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];
			if (nx < 0 || nx >= r || ny < 0 || ny >= c || visitedSwan[nx][ny]) continue;
			visitedSwan[nx][ny] = 1;
			if (a[nx][ny] == '.') swanQ.push({ nx, ny });
			else if (a[nx][ny] == 'X') swanTempQ.push({ nx, ny });
			else if (a[nx][ny] == 'L') return true;
		}
	}
	return false;
}

void waterMelt() {
	while (waterQ.size()) {
		int x = waterQ.front().first;
		int y = waterQ.front().second;
		waterQ.pop();

		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];
			if (nx < 0 || nx >= r || ny < 0 || ny >= c || visitedWater[nx][ny]) continue;
			if (a[nx][ny] == 'X') {
				visitedWater[nx][ny] = 1;
				waterTempQ.push({ nx, ny });
				a[nx][ny] = '.';
			}
		}
	}
}

int main() {
	ios_base::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);

	cin >> r >> c;
	for (int i = 0; i < r; i++) {
		for (int j = 0; j < c; j++) {
			cin >> a[i][j];

			if (a[i][j] == 'L') swan = { i, j };// 백조
			if (a[i][j] == 'L' || a[i][j] == '.') visitedWater[i][j] = 1, waterQ.push({ i, j });// 백조 || 물
		}
	}

	swanQ.push({ swan.first, swan.second });
	visitedSwan[swan.first][swan.second] = 1;
	while (true) {
		if (moveSwan()) break;
		waterMelt();
		waterQ = waterTempQ;
		swanQ = swanTempQ;
		Qclear(waterTempQ);
		Qclear(swanTempQ);
		day++;
	}
	cout << day << "\n";
	return 0;
}

File: 백준/Platinum/3197. 백조의 호수/백조의 호수.java
import java.io.*;
import java.util.*;

public class Main {
    static final int[] dx = {-1, 0, 1, 0};
    static final int[] dy = {0, 1, 0, -1};
    static int r, c, swanX, swanY, day;
    static int[][] a;
    static boolean[][] visitedSwan, visitedWater;
    static Queue<Data> swanQ = new ArrayDeque<>(), swanTempQ = new ArrayDeque<>(), waterQ = new ArrayDeque<>(), waterTempQ = new ArrayDeque<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        a = new int[r][c];
        visitedSwan = new boolean[r][c];
        visitedWater = new boolean[r][c];
        for (int i = 0; i < r; i++) {
            String s = br.readLine();
            for (int j = 0; j < c; j++) {
                if (s.charAt(j) == '.') { // 물
                    visitedWater[i][j] = true;
                    waterQ.offer(new Data(i, j));
                    a[i][j] = 0;
                }
                if (s.charAt(j) == 'L') { // 백조(물)
                    swanX = i;
                    swanY = j;
                    a[i][j] = 2;
                    visitedWater[i][j] = true;
                    waterQ.offer(new Data(i, j));
                } else if (s.charAt(j) == 'X') { // 빙판 공간
                    a[i][j] = 1;
                }
            }
        }
        visitedSwan[swanX][swanY] = true;
        swanQ.offer(new Data(swanX, swanY));
        while (true) {
            if (moveSwan()) break;
            waterMelt();
            swanQ = swanTempQ;
            waterQ = waterTempQ;
            swanTempQ = new ArrayDeque<>();
            waterTempQ = new ArrayDeque<>();
            day++;
        }
        System.out.println(day);
    }

    static void waterMelt() {
        while (!waterQ.isEmpty()) {
            Data now = waterQ.poll();
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if (nx < 0 || nx >= r || ny < 0 || ny >= c || visitedWater[nx][ny]) continue; // 범위 || 방문
                if (a[nx][ny] == 1) {
                    visitedWater[nx][ny] = true;
                    waterTempQ.offer(new Data(nx, ny));
                    a[nx][ny] = 0;
                }
            }
        }
    }

    static boolean moveSwan() {
        while (!swanQ.isEmpty()) {
            Data now = swanQ.poll();
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if (nx < 0 || nx >= r || ny < 0 || ny >= c || visitedSwan[nx][ny]) continue; // 범위 || 방문
                visitedSwan[nx][ny] = true;
                if (a[nx][ny] == 0) swanQ.offer(new Data(nx, ny));
                else if (a[nx][ny] == 1) swanTempQ.offer(new Data(nx, ny));
                else if (a[nx][ny] == 2) return true;
            }
        }
        return false;
    }

    static class Data {
        int x, y;

        public Data(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}